hbase-server/src/main/java/org/apache/hadoop/hbase/client/AsyncClusterConnection.java

```diff
@@ -63,6 +64,13 @@
   CompletableFuture<Long> replay(TableName tableName, byte[] encodedRegionName, byte[] row,
       List<Entry> entries, int replicaId, int numRetries, long operationTimeoutNs);
 
+  /**
+   * Replicate wal edits for replica regions, the location for the replica is already included.
+   * The return value is the edits we skipped, as the original return value is useless.
+   */
+  CompletableFuture<Long> replay(TableName tableName, byte[] encodedRegionName, List<Entry> entries,
+    int replicaId, int numRetries, long operationTimeoutNs, HRegionLocation loc);
```

Comment:
Yeah, thought about it. It is messy with my current implementation. The whole idea is that since shipper only ships to one replica region sever and passing in location will save extra cost to look up location. 
For the case of region move case, the retry does not  work in this implementation as location is fixed. I think it will be better just let vanilla implementation to take care of it. Let me clean this up and update the patch.